crypto/bio/bss_conn.c

```diff
@@ -188,7 +188,16 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
 
         case BIO_CONN_S_BLOCKED_CONNECT:
             i = BIO_sock_error(b->num);
-            if (i) {
+            if (i != 0) {
+                if ((c->addr_iter = BIO_ADDRINFO_next(c->addr_iter)) != NULL) {
+                    /*
+                     * if there are more addresses to try, do that first
+                     */
+                    BIO_closesocket(b->num);
+                    c->state = BIO_CONN_S_CREATE_SOCKET;
+                    ERR_clear_error();
+                    break;
+                }
                 BIO_clear_retry_flags(b);
```

Comment:
In general I think this makes sense, if the blocking connect fails the
error handling is here:
https://github.com/openssl/openssl/blob/c74aaa3920f116fe4c1003153838144c37d6e527/crypto/bio/bss_conn.c#L167-L175
but if the non-blocking connect does not fail immediately
the error is in the SO_ERROR and we end up here.
Not sure if the call to BIO_clear_retry_flags should move up, since
the connect was retried and failed.  This flag might be visible in the callback.

Reply:
Thanks @bernd-edlinger for having a look at this and commenting.
When I debugged this I found that for non-blocking BIOs first `BIO_CONN_S_CONNECT` is reached and fails, and on the first retry (done by `BIO_connect_retry()` `BIO_CONN_S_BLOCKED_CONNECT` is reached.
I don't understand why this complicated retry is needed at all, but it works.

Reply:
Yeah, a name resolution can yield several addresses, this tries each address,
until the first succeeds.
In my code when I write a server I usually let the user decide on IPv4 or IPv6 first,
and then try in that sequence for instance.  I think we have added s_client option:
```
 -4                         Use IPv4 only
 -6                         Use IPv6 only
```
does that work for you?

Reply:
I certainly could add such options to the CMP client, but I see not much motivation to do so. And the corrected retry behavior may be needed also for other reasons, not only IP v4 vs. v6.

Reply:
There must be a way to set connect_family to BIO_FAMILY_IPV4
BIO_C_SET_CONNECT: /num == 3
this is rotten bad code without any comments.

Reply:
BIO_set_conn_ip_family(b,BIO_FAMILY_IPV4)

Reply:
I can only say, that it is really worth it to give the user a way to prefer
IPv4 or IPv6, not only in this project.
Of course lots of other things can go wrong but then someone else
gets the blame...

Reply:
Of course for apps like `s_client` focusing on network connections it makes much sense to have `-4` and `-6` options, while for the specific CMP client use case I'm still a bit reluctant to add such rather network-special options, in particular since this app already has tons of options.

On the other hand, the four HTTP client functions declared in `include/openssl/http.h`:
```
BIO *OSSL_HTTP_get(const char *url, const char *proxy, const char *no_proxy,
                   BIO *bio, BIO *rbio,
                   OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
                   const STACK_OF(CONF_VALUE) *headers,
                   int maxline, unsigned long max_resp_len, int timeout,
                   const char *expected_content_type, int expect_asn1);
ASN1_VALUE *OSSL_HTTP_get_asn1(const char *url,
                               const char *proxy, const char *no_proxy,
                               BIO *bio, BIO *rbio,
                               OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
                               const STACK_OF(CONF_VALUE) *headers,
                               int maxline, unsigned long max_resp_len,
                               int timeout, const char *expected_content_type,
                               const ASN1_ITEM *it);
ASN1_VALUE *OSSL_HTTP_post_asn1(const char *server, const char *port,
                                const char *path, int use_ssl,
                                const char *proxy, const char *no_proxy,
                                BIO *bio, BIO *rbio,
                                OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
                                const STACK_OF(CONF_VALUE) *headers,
                                const char *content_type,
                                const ASN1_VALUE *req, const ASN1_ITEM *req_it,
                                int maxline, unsigned long max_resp_len,
                                int timeout, const char *expected_ct,
                                const ASN1_ITEM *rsp_it);
BIO *OSSL_HTTP_transfer(const char *server, const char *port, const char *path,
                        int use_ssl, const char *proxy, const char *no_proxy,
                        BIO *bio, BIO *rbio,
                        OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
                        const STACK_OF(CONF_VALUE) *headers,
                        const char *content_type, BIO *req_mem,
                        int maxline, unsigned long max_resp_len, int timeout,
                        const char *expected_ct, int expect_asn1,
                        char **redirection_url);
```
are meant to be as generally applicable as possible.
Shall I add to those an `int family` parameter?

Reply:
It is just my experience, with OPC/UA Clients that it will probably be helpful.
But when I look at those parameters, they are easy to confuse, 90% of the paramters
are just how to establish a tcp connection, but what about udp?
Maybe also add a protocol parameter?
And  put all the common how-to-connect parameters into a structure, maybe
with a parser, that parses a connection string and fill in all the params?

Reply:
but also please as another PR.

Reply:
Yep, will do as soon as I have more bandwidth again.
I also was thinking that the many function parameter need to be abstracted/structured.